ui-admin/src/portal/publish/PortalEnvDiffProvider.tsx

```diff
@@ -0,0 +1,61 @@
+import { useNavigate, useParams } from 'react-router-dom'
+import React, { useEffect, useState } from 'react'
+import Api, { PortalEnvironmentChange } from 'api/api'
+import { Store } from 'react-notifications-component'
+import { failureNotification, successNotification } from 'util/notifications'
+import _cloneDeep from 'lodash/cloneDeep'
+import LoadingSpinner from 'util/LoadingSpinner'
+import PortalEnvDiffView from './PortalEnvDiffView'
+import { Portal, PortalEnvironment } from '@juniper/ui-core/build/types/portal'
+
+type PortalEnvDiffProviderProps = {
+  portal: Portal,
+  portalEnv: PortalEnvironment,
+  updatePortal: (portal: Portal) => void
+}
+/**
+ * loads a diff between two environments, based on the passed-in environment and an environment name in a url param
+ * also contains logic for updating an environment with a changeset
+ */
+const PortalEnvDiffProvider = ({ portal, portalEnv, updatePortal }: PortalEnvDiffProviderProps) => {
+  const params = useParams()
+  const sourceEnvName: string | undefined = params.sourceEnvName
+  const [isLoading, setIsLoading] = useState(true)
+  const [diffResult, setDiffResult] = useState<PortalEnvironmentChange | null>(null)
```

Comment:
I like using null instead of undefined in cases like this because it makes it a bit easier to see what's going on in the debugger (if I see something is `null` I know it's at least being set somewhere).  But it's just a small personal preference.  If React leans the other way, we should go with that